design/tg-branches.md

```diff
@@ -0,0 +1,111 @@
+1. Which environment(s) do PR TestGrid runs execute against?
+1. How to get a Kurl installer for packages that aren't yet supported (i.e. break validation)?
+1. How navigate addons being hard-coded in tools and scripts?
+
+## High-Level Design
+
+Use the following environments:
+1. Use Prod TestGrid for executing PR tests
+1. Use staging.kurl.sh for installers
+
+Code changes that are required:
+1. The Kurl installer spec will be updated to allow installers to be generated with `upsteam_override`.
+1. kurl.sh will be modified to allow any version for an addon in the spec with the `upstream_override` enabled.
+1. `tgrun` will be updated
+   1. Since this proposal will only really test updated addons, a flag to run only a single "latest" kurl installer against all OSes.
+   1. Flags to specify addon version and download url
+1. Modify install.sh to use `upstream_override` bucket path from the kurl spec.
+   1. Possibly modify `yamltobash` utils (and others?)
+1. Add a github action to run on a PR creation/update.
+   1. Do a diff and see if any of the addon dirs have been modified.
+   1. Push packages to `<s3>/pr` for modified dirs (wait?)
+   1. `tgrun queue --ref=<short commit>-pr-<#> --latest-only --version=contour=1.12.0 --upstream_override=<s3 url in previous step>`
+   1. (Optional) Poll results and/or add TestGrid url in PR.
+
+### Limitations
+1. This process won't work for new addons or changes to the actual installer scripting.
+    1. the kurl.sh script and `yamltobash` need to be changed here and served somewhere that can be access for each test instance config.
+1. Can't use expect 'latest' test configuration to resolve to the newest config
+    1. Latest is interpretted by kurl.sh api. Could get around this with some magic in kru
+   
+
+## Detailed Design
+
+Updated Kurl Installer Spec to get around validation on kurl.sh (Courtesy @lavery)a:
+```yaml
+apiVersion: "cluster.kurl.sh/v1beta1"
+kind: "Installer"
+metadata: 
+  name: "latest"
+spec: 
+  kubernetes: 
+    version: "latest"
+  weave: 
+    version: "latest"
+  rook: 
+    version: "latest"
+  contour: 
+    version: "latest"
+    upstream_override: "<bucket-url>/pr/<commit hash>-contour-1.11.0.tar.gz"
+  longhorn:
+    version: "latest"
+    upstream_override: "<bucket-url>/pr/<commit hash>-contour-1.11.0.tar.gz"
+  ...
+```
+
+## Testing
+
+Initial testing will be challenging considering it requires changes to a lot of the components. 
+Proposing to break this up into these PRs to mitigate risk and manage code review:
+- Change kurl.sh web to adopt new kurl spec, run test suite.
+- Modify `tgrun` with proposed changes, test that kurl.sh install fail for the new spec format. 
+- Thorough manual testing for changes to install.sh that pull from the repository.
+- Attempt an addon upgrade and see that the GH action produces the desired affect.
+
+Regressions should be detected  quickly as this will run as part of the PR process on the repo. 
+
+## Alternatives Considered
```

Comment:
Could it be possible to specify an installer ID to run for a given PR? Perhaps by using a convention of branch naming where you can deduce the installer ID? This could get around addons that aren't in `latest`, as well as helping with new addons.

Reply:
If I'm understanding how everything works, I think you could easily make a `tgrun --installer-id=<id>` flag to test a specific installer against all OSes, which could be independently valuable in the future for testing an arbitrary installer instead of the just the ones coded into `tgrun`. 

Not sure this would help with PRs though. To have an installer ID, the web api has to receive a spec. So someone would need to submit the spec manually (defeats the purpose), or you generate the ID based on a spec sent automatically as part of the PR (the current proposal using `tgrun`).

New addons will always be a problem because install.sh needs to be updated to handle the new addon, but can't be rendered until a new version of the web api is pushed.